compiler/rustc_trait_selection/src/traits/wf.rs

```diff
@@ -168,6 +168,23 @@ pub fn predicate_obligations<'tcx>(
     wf.normalize(infcx)
 }
 
+/// Whether we should emit nested obligations containing `value`.
+/// This is a hack to deal with higher ranked types as we require implications
+/// in binders for a more correct implementation.
+///
+/// `WF(for<'a> fn(&'a T))` would otherwise emit a `T: 'a` bound for `&'a T`
+/// which would require `T: 'static`. What we actually want here is the type
+/// `for<'a> where { T: 'a } fn(&'a T)`, at which point the `T: 'a` bound is
+/// trivially implied from the `param_env`.
+///
+/// This means that we WF bounds aren't always sufficiently checked for now.
+fn emit_obligations_for<'tcx, T: TypeVisitable<'tcx>>(value: T) -> bool {
```

Comment:
Maybe make this a method on TypeVisitable and add warning doc comments on has_escaping_bound_var